### PR TITLE
EVG-7843 Make generate.tasks races less likely

### PR DESCRIPTION
**Pull request by bsamek**

### CODE REVIEW

File: model/generate.go [hunk 1]
```diff
@@ -118,57 +117,46 @@ func ParseProjectFromJSON(data []byte) (GeneratedProject, error) {
...
-		return nil, nil, nil, nil, nil,
-			gimlet.ErrorResponse{StatusCode: http.StatusBadRequest, Message: fmt.Sprintf("unable to find version %s", t.Version)}
-	}
-	p, pp, err := LoadProjectForVersion(v, t.Project, false)
-	if err != nil {
-		return nil, nil, nil, nil, nil,
-			gimlet.ErrorResponse{StatusCode: http.StatusBadRequest, Message: errors.Wrapf(err, "error getting project for version %s", t.Version).Error()}
-	}
-
+func (g *GeneratedProject) NewVersion(p *Project, pp *ParserProject, v *Version) (*Project, *ParserProject, *Version, *projectMaps, error) {
```
Comment (tychoish) [inline model/generate.go:120]: why do we have a public method returning a private type that isn't wrapped in an interface? 

is the *Project that's returned the same as the *Project that's passed in? can we just mutate the pointer that we're handed?

File: model/generate.go [hunk 2]
```diff
@@ -118,57 +117,46 @@ func ParseProjectFromJSON(data []byte) (GeneratedProject, error) {
...
 	if err != nil {
-		return nil, nil, nil, nil, nil,
-			gimlet.ErrorResponse{StatusCode: http.StatusBadRequest, Message: errors.Wrap(err, "error translating project").Error()}
+		return nil, nil, nil, nil, errors.Wrap(err, "error translating project")
 	}
-	return p, newPP, v, t, &cachedProject, nil
+	return p, newPP, v, &cachedProject, nil
 }
 
 func (g *GeneratedProject) Save(ctx context.Context, p *Project, pp *ParserProject, v *Version, t *task.Task, pm *projectMaps) error {
```
Comment (tychoish) [inline model/generate.go:142]: public function with a private type as an argument is awkward.

File: model/generate.go [hunk 3]
```diff
@@ -118,57 +117,46 @@ func ParseProjectFromJSON(data []byte) (GeneratedProject, error) {
...
+	if t == nil {
+		return errors.Errorf("unable to find task %s", g.TaskID)
+	}
+	if t.GeneratedTasks {
+		grip.Debug(message.Fields{
+			"message": "skipping attempting to update parser project because another generator marked the task complete",
+			"task":    t.Id,
+			"version": t.Version,
+		})
+		return mongo.ErrNoDocuments
```
Comment (tychoish) [inline model/generate.go:157]: o.O should we just return nil?

File: units/generate_tasks.go [hunk 4]
```diff
@@ -76,8 +76,40 @@ func (j *generateTasksJob) generate(ctx context.Context, t *task.Task) error {
 		return nil
 	}
 
+	v, err := model.VersionFindOneId(t.Version)
```
Comment (tychoish) [inline units/generate_tasks.go:79]: shouldn't this all be in the models package?

File: units/generate_tasks.go [hunk 5]
```diff
@@ -170,6 +202,26 @@ func (j *generateTasksJob) generate(ctx context.Context, t *task.Task) error {
...
+	if t == nil {
+		return errors.Errorf("unable to find task %s", j.TaskID)
+	}
+	if t.GeneratedTasks {
+		grip.Debug(message.Fields{
+			"message": "handledError encountered task that already generating, nooping",
+			"task":    t.Id,
+			"version": t.Version,
+		})
+		return mongo.ErrNoDocuments
```
Comment (tychoish) [inline units/generate_tasks.go:222]: we should just return nil here, probably?

File: model/generate.go (hunk 1 above)
Comment (bsamek) [reply model/generate.go:120]: What would you gain from not returning them? The underlying machinery doesn't mutate. You'd be generating a new thing and then assigning the pointer to the new thing. I think it's clearer to do it this way given that we're creating something new. Nothing has changed. We've just reduced the amount of work this function is doing, so now it takes an argument.

File: model/generate.go (hunk 2 above)
Comment (bsamek) [reply model/generate.go:142]: 100% yes, but I don't think refactoring this should be within the scope of this ticket.

File: model/generate.go (hunk 3 above)
Comment (bsamek) [reply model/generate.go:157]: No. The callers of this function rely on `mongo.ErrNoDocuments` to signify a race. Again, we can refactor, but not within this ticket, which has a different goal. I can open a ticket for a refactor.

File: units/generate_tasks.go (hunk 4 above)
Comment (bsamek) [reply units/generate_tasks.go:79]: Yeah, this is mostly just calling a function from the model package and then doing some error checking.

File: units/generate_tasks.go (hunk 5 above)
Comment (bsamek) [reply units/generate_tasks.go:222]: No, the callers rely on this to signify a race.

**Comment by bsamek** [comment]
> Besides the NewVersion function signature, my only thought is that for some t.GeneratedTasks we return nil, and for some we return mongo.ErrNoDocument. We should standardize this, because I think sometimes we'll print "generate.tasks finished" and other times "generate.tasks noop". But logically everything looks good.

@ablack12 Ah, true. Fixed.

**Comment by ablack12** [comment]
LGTM 👍

**Pull request merged**